First implementation of NL Loki-scc variant on GPU

piotrows opened the pull request:
This PR brings in working implementation of cloudsc2-nl loki-scc variant on GPU.  loki-scc-hoist variant is not functional, although it can be compiled with a manual edit of the Loki-output of the driver. When run, it delivers extremely poor performance - possibly input on the side of Loki transformation is needed to improve.

piotrows committed:
loki-idem works with nvhpc, scc-hoist compilation fails because of empty acc enter data create, scc fails on runtimne with Failing in Thread:1
call to cuStreamSynchronize returned error 700: Illegal address during kernel execution

piotrows committed:
Weird bug connected with CETA, two files that actually pass cloudsc2 in scc variant are attached, signaling FPE in diagnostics. CETA needed in cloudsc2 arguments even tough it is not later accessed, as replaced by assosciate with ycld%ceta. prints do not matter. acc parallel loop  gang must be degraded to acc parallel, as in the example scc files

piotrows committed:
scc examples removed as they interfere with compilation, bugfix to allocate local_ycld%CETA. Still doesnt work

piotrows committed:
nl loki-scc now working with CETA workaround, scc-hoist requires manual edit and yields poor performance

piotrows committed:
Main README.md update on Loki variants

mlange05 committed:
Merge branch 'develop' into napz-loki-nl-gpu

mlange05 committed:
Loki: Re-enable OpenACC generation for Loki-SCC/SCC-H

mlange05 committed:
Loki-NL: Fix `!$loki data` pragma region to correctly honour timers

mlange05 commented on src/cloudsc2_nl_loki/cloudsc_driver_loki_mod.F90:
```diff
@@ -72,22 +73,30 @@ SUBROUTINE CLOUDSC_DRIVER( &
-    TYPE(TEPHLI)    :: YPHLI
-    TYPE(TECLD)     :: YCLD
-    TYPE(TECLDP)    :: YCLDP
+    TYPE(TOMCST)    :: YDCST, LOCAL_YDCST
```
The argument objects have no intent!

mlange05 commented on src/cloudsc2_nl_loki/cloudsc_driver_loki_mod.F90:
```diff
@@ -72,22 +73,30 @@ SUBROUTINE CLOUDSC_DRIVER( &
     !   write(0,1003) NUMPROC,NUMOMP,NGPTOTG,NPROMA,NGPBLKS
     ! end if
 
+    LOCAL_YDCST=YDCST
```
I'm not sure that this is required here. Latest NVHPC toolchain should happily accept the argument object without needing a local copy.

piotrows replied:
It is needed, I've checked it.

mlange05 replied:
I think I managed to fir this; declaration of the type objects was off. Please have a look and check this works.

mlange05 commented on src/cloudsc2_nl_loki/dwarf_cloudsc.F90:
```diff
@@ -101,6 +102,7 @@ PROGRAM DWARF_CLOUDSC
 ! Instead we do this similar computation involving values already available
 DO JK=1,GLOBAL_STATE%KLEV
   YRECLD%CETA(JK)= GLOBAL_STATE%PAP(1,JK,1)/GLOBAL_STATE%PAPH(1,GLOBAL_STATE%KLEV+1,1)
+  LCETA(JK)=YRECLD%CETA(JK)
```
I think this is not needed. We've encountered this in other contexts, where the problem is that ALLOCATABLE components of a derived type object are not copied to device with the rest of the components unless explicitly forced via `!$acc create copyin` or `!$acc data` regions. Since Loki's insertion of the OpenACC offload regions accepts hand-rolled pragmas in addition to the auto-generated ones, we can probably add the missing pragma in the driver directly, so we don;t have to do all this back-bending here.

FWIW, this will hopefully fixed in future Loki-SCC versions, where Loki will add the missing copies explicitly, but that is future work for now.

mlange05 replied:
I can have a quick go at this and see how to best handle this. I'll ping once I can say more...

mlange05 commented on src/common/CMakeLists.txt:
```diff
@@ -58,5 +58,69 @@ ecbuild_add_library( TARGET cloudsc2-common-lib
+
+
+
+
```
Debug leftover, should be removed!

mlange05 commented on src/cloudsc2_nl_loki/README.md:
```diff
@@ -0,0 +1,4 @@
+Current implementation provides working idem and scc variants, the latter with 3.5 GFlops/s. scc-hoist currently requires manual edit of 
```
Don't think this is needed anymore. Instead maybe update the central README with latest state?

mlange05 commented on bundle.yml:
```diff
@@ -50,6 +50,14 @@ options :
+    - with-gpu :
+        help  : Enable GPU kernels
+        cmake : >
+            ENABLE_CLOUDSC_GPU_SCC=ON
```
I am pretty sure none of these have any effect, as they are not defined in the CMakeLists.txt of the Loki variant. In fact, since we have no non-loki GPU configurations in this dwarf, this flag here seems redundant.

Having said that, the fact that the GPU compilation toolchain "just picks up" GPU compiler options is not ideal. However, properly configuring the GPU-specific build options should probably happen in a separate follow-up PR. For now we can probably just remove this?

mlange05 commented on README.md:
````diff
@@ -39,9 +39,15 @@ install the bundle via:
+ ./cloudsc-bundle build --with-gpu --with-loki --loki-frontend=fp --arch=./arch/ecmwf/hpc2020/nvhpc/22.1 
+```
+NVHPC will not build Loki variant with --build-type=debug due to the compiler error. Possibly, look into the compiler flags is needed here.
+
````
We'll also need updated run instructions here, in particular for Loki-SCC where we need `NV_ACC_CUDA_HEAPSIZE=9GB` or similar.

mlange05 reviewed (changes requested):
Ok, this is starting to shape up nicely. I've pushed a few small changes that make this functional and correct already, but I've also flagged a few places where we need to clean up the changes a little further. 

In particular, I don't think the local copy trick for CETA is needed, as this is a common issue that we face in EC-physics often: for ALLOCATABLE derived type members we need explicit offload pragmas for the components, which Loki does not yet generated correctly. I'll have a look at fixing this myself, and ping back here when I know a little more. 

@piotrows In the mean time, can you please take care of some of the other niggles I've flagged, and verify that this now runs for you as expected?

mlange05 commented on src/cloudsc2_nl_loki/CMakeLists.txt:
```diff
@@ -119,7 +119,7 @@ if( HAVE_CLOUDSC2_NL_LOKI )
             loki-scc-hoist/cloudsc2.scc_hoist.F90
             loki-scc-hoist/cloudsc_driver_loki_mod.scc_hoist.F90
-	    DEPENDS satur.F90 cloudsc2.F90 cloudsc_driver_loki_mod.F90 
+	      DEPENDS satur.F90 cloudsc2.F90 cloudsc_driver_loki_mod.F90 
```
Rogue tab spacing? Please only use spaces, no tabs, in cmake files.

piotrows committed:
Cleanup of unnecessary bundle options, spurious tabs, delete spurious Readme, remove CMaKeLists debug leftovers, update main README on comulsory CUDA heapsize set.   #

piotrows committed:
Remove unused STATE_TYPE arguments to driver in NL-Loki, add INTENT to derived type arguments

piotrows committed:
Add initialization to BUFFER_LOC to avoid non-deterministic errors under GCC

mlange05 committed:
CLOUDSC2-NL: Promote ZQSAT to buffer data between kernel calls

mlange05 committed:
Common: Make YRECLD ALLOCATABLE, remove "declare copyin" for YRECDLP

mlange05 committed:
Loki-NL: Remove use of local copies for parameter types

mlange05 committed:
README: Small changes to the run instructions for Loki GPU versions

mlange05 commented on README.md:
````diff
@@ -39,9 +39,26 @@ install the bundle via:
 ```
+Targetting GPU, one should use:
+```
+ ./cloudsc-bundle build --with-gpu --with-loki --loki-frontend=fp --arch=./arch/ecmwf/hpc2020/nvhpc/22.1 
````
Uhhh, I think `--with-gpu` has been removed, no?

piotrows commented:
I've checked and the code seem to work well, but I see two issues to resolve:
1) The scc variants won't compile in build type debug on GPU seem to be caused by the OpenMP flags, of no interest at the moment in cloudsc2. The solution for ATOS seems to be to comment out from toolchain.cmake: OpenMP Flags AND "-mp gpu" flag from the OpenACC Flags. In turn, deleting just the latter seem sufficient on 23.1 on Juwels.
2) The relevant section of README should include information on turning off the acc directives via cmake, i.e.:
On ATOS HPC, Loki variant may be built targeting CPU using
```
./cloudsc-bundle build --clean --with-loki --loki-frontend=fp -cmake=ENABLE_ACC=off --arch=./arch/ecmwf/hpc2020/gnu/11.2.0/
./cloudsc-bundle build --clean --with-loki --loki-frontend=fp --arch=./arch/ecmwf/hpc2020/intel/2021.4.0
```
The line mentioning the build error when in debug should be consequently deleted if we go the 1) route.

mlange05 committed:
NVHPC: Remove unsupported OpenMP-offload flags from NVHPC toolchain

This causes issues in debug build and is currently untested and
thus unsupported.

mlange05 committed:
README: Make Loki build and run instructions more specific

mlange05 commented:
> I've checked and the code seem to work well, but I see two issues to resolve:
> 
> 1. The scc variants won't compile in build type debug on GPU seem to be caused by the OpenMP flags, of no interest at the moment in cloudsc2. The solution for ATOS seems to be to comment out from toolchain.cmake: OpenMP Flags AND "-mp gpu" flag from the OpenACC Flags. In turn, deleting just the latter seem sufficient on 23.1 on Juwels.
> 2. The relevant section of README should include information on turning off the acc directives via cmake, i.e.:
>    On ATOS HPC, Loki variant may be built targeting CPU using
> 
> ```
> ./cloudsc-bundle build --clean --with-loki --loki-frontend=fp -cmake=ENABLE_ACC=off --arch=./arch/ecmwf/hpc2020/gnu/11.2.0/
> ./cloudsc-bundle build --clean --with-loki --loki-frontend=fp --arch=./arch/ecmwf/hpc2020/intel/2021.4.0
> ```
> 
> The line mentioning the build error when in debug should be consequently deleted if we go the 1) route.

Ok, thanks for the pointers. I've removed the unsupported OpenMP-offload remnants and cleaned up the Loki/GPU section a little, to make it clear that the Loki variants are really just for GPU usage. I've left the instructions to build non-OpenACC variants, and tested them locally. I've also remove the performance section, as this was haphazard at best, and I'd like to wait until we've done a full write-up of all performance numbers for the dwarf.

@piotrows Please have another look, and indicate here if you're happy with merging this.

mlange05 reviewed (approved):
Following CMake sanitisation and README update, I think this is GTG now.

piotrows committed:
Fixed a few typos in README.md
